File: spark/src/main/scala/org/apache/spark/sql/delta/LastCheckpointInfo.scala
/*
 * Copyright (2021) The Delta Lake Project Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.spark.sql.delta

import scala.collection.mutable.ArrayBuffer

import org.apache.spark.sql.delta.actions.{CheckpointMetadata, SidecarFile, SingleAction}
import org.apache.spark.sql.delta.sources.DeltaSQLConf
import org.apache.spark.sql.delta.util.FileNames.{checkpointVersion, numCheckpointParts}
import org.apache.spark.sql.delta.util.JsonUtils
import com.fasterxml.jackson.annotation.{JsonIgnore, JsonPropertyOrder}
import com.fasterxml.jackson.databind.{DeserializationFeature, JsonNode}
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.fasterxml.jackson.databind.node.ObjectNode
import org.apache.commons.codec.digest.DigestUtils
import org.apache.hadoop.fs.FileStatus

import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.types.StructType

/**
 * Information about the V2 Checkpoint in the LAST_CHECKPOINT file
 * @param path             file name corresponding to the uuid-named v2 checkpoint
 * @param sizeInBytes      size in bytes for the uuid-named v2 checkpoint
 * @param modificationTime modification time for the uuid-named v2 checkpoint
 * @param nonFileActions   all non file actions for the v2 checkpoint. This info may or may not be
 *                         available. A None value means that info is missing.
 *                         If it is not None, then it should have all the non-FileAction
 *                         corresponding to the checkpoint.
 * @param sidecarFiles     sidecar files corresponding to the v2 checkpoint. This info may or may
 *                         not be available. A None value means that this info is missing.
 *                         An empty list denotes that the v2 checkpoint has no sidecars.
 */
case class LastCheckpointV2(
    path: String,
    sizeInBytes: Long,
    modificationTime: Long,
    nonFileActions: Option[Seq[SingleAction]],
    sidecarFiles: Option[Seq[SidecarFile]]) {

  @JsonIgnore
  lazy val checkpointMetadataOpt: Option[CheckpointMetadata] =
    nonFileActions.flatMap(_.map(_.unwrap).collectFirst { case cm: CheckpointMetadata => cm })

}

object LastCheckpointV2 {
  def apply(
      fileStatus: FileStatus,
      nonFileActions: Option[Seq[SingleAction]] = None,
      sidecarFiles: Option[Seq[SidecarFile]] = None): LastCheckpointV2 = {
    LastCheckpointV2(
      path = fileStatus.getPath.getName,
      sizeInBytes = fileStatus.getLen,
      modificationTime = fileStatus.getModificationTime,
      nonFileActions = nonFileActions,
      sidecarFiles = sidecarFiles)
  }
}

/**
 * Records information about a checkpoint.
 *
 * This class provides the checksum validation logic, needed to ensure that content of
 * LAST_CHECKPOINT file points to a valid json. The readers might read some part from old file and
 * some part from the new file (if the file is read across multiple requests). In some rare
 * scenarios, the split read might produce a valid json and readers will be able to parse it and
 * convert it into a [[LastCheckpointInfo]] object that contains invalid data. In order to prevent
 * using it, we do a checksum match on the read json to validate that it is consistent.
 *
 * For old Delta versions, which do not have checksum logic, we want to make sure that the old
 * fields (i.e. version, size, parts) are together in the beginning of last_checkpoint json. All
 * these fields together are less than 50 bytes, so even in split read scenario, we want to make
 * sure that old delta readers which do not do have checksum validation logic, gets all 3 fields
 * from one read request. For this reason, we use `JsonPropertyOrder` to force them in the beginning
 * together.
 *
 * @param version the version of this checkpoint
 * @param size the number of actions in the checkpoint, -1 if the information is unavailable.
 * @param parts the number of parts when the checkpoint has multiple parts. None if this is a
 *              singular checkpoint
 * @param sizeInBytes the number of bytes of the checkpoint
 * @param numOfAddFiles the number of AddFile actions in the checkpoint
 * @param checkpointSchema the schema of the underlying checkpoint files
 * @param checksum the checksum of the [[LastCheckpointInfo]].
 */
@JsonPropertyOrder(Array("version", "size", "parts"))
case class LastCheckpointInfo(
    version: Long,
    size: Long,
    parts: Option[Int],
    @JsonDeserialize(contentAs = classOf[java.lang.Long])
    sizeInBytes: Option[Long],
    @JsonDeserialize(contentAs = classOf[java.lang.Long])
    numOfAddFiles: Option[Long],
    checkpointSchema: Option[StructType],
    v2Checkpoint: Option[LastCheckpointV2] = None,
    checksum: Option[String] = None) {

  @JsonIgnore
  def getFormatEnum(): CheckpointInstance.Format = parts match {
    case _ if v2Checkpoint.nonEmpty => CheckpointInstance.Format.V2
    case Some(_) => CheckpointInstance.Format.WITH_PARTS
    case None => CheckpointInstance.Format.SINGLE
  }

  /** Whether two [[LastCheckpointInfo]] represents the same checkpoint */
  def semanticEquals(other: LastCheckpointInfo): Boolean = {
    CheckpointInstance(this) == CheckpointInstance(other)
  }
}

object LastCheckpointInfo {

  val STORED_CHECKSUM_KEY = "checksum"

  /** Whether to store checksum OR do checksum validations around [[LastCheckpointInfo]]  */
  def checksumEnabled(spark: SparkSession): Boolean =
    spark.sessionState.conf.getConf(DeltaSQLConf.LAST_CHECKPOINT_CHECKSUM_ENABLED)

  /**
   * Returns the json representation of this [[LastCheckpointInfo]] object.
   * Also adds the checksum to the returned json if `addChecksum` is set. The checksum can be
   * used by readers to validate consistency of the [[LastCheckpointInfo]].
   * It is calculated using rules mentioned in "JSON checksum" section in PROTOCOL.md.
   */
  def serializeToJson(
      lastCheckpointInfo: LastCheckpointInfo,
      addChecksum: Boolean,
      suppressOptionalFields: Boolean = false): String = {
    if (suppressOptionalFields) {
      return JsonUtils.toJson(
        LastCheckpointInfo(
          lastCheckpointInfo.version,
          lastCheckpointInfo.size,
          lastCheckpointInfo.parts,
          sizeInBytes = None,
          numOfAddFiles = None,
          v2Checkpoint = None,
          checkpointSchema = None))
    }

    val jsonStr: String = JsonUtils.toJson(lastCheckpointInfo.copy(checksum = None))
    if (!addChecksum) return jsonStr
    val rootNode = JsonUtils.mapper.readValue(jsonStr, classOf[ObjectNode])
    val checksum = treeNodeToChecksum(rootNode)
    rootNode.put(STORED_CHECKSUM_KEY, checksum).toString
  }

  /**
   * Converts the given `jsonStr` into a [[LastCheckpointInfo]] object.
   * if `validate` is set, then it also validates the consistency of the json:
   *  - calculating the checksum and comparing it with the `storedChecksum`.
   *  - json should not have any duplicates.
   */
  def deserializeFromJson(jsonStr: String, validate: Boolean): LastCheckpointInfo = {
    if (validate) {
      val (storedChecksumOpt, actualChecksum) = LastCheckpointInfo.getChecksums(jsonStr)
      storedChecksumOpt.filter(_ != actualChecksum).foreach { storedChecksum =>
        throw new IllegalStateException(s"Checksum validation failed for json: $jsonStr,\n" +
          s"storedChecksum:$storedChecksum, actualChecksum:$actualChecksum")
      }
    }

    // This means:
    // 1) EITHER: Checksum validation is config-disabled
    // 2) OR: The json lacked a checksum (e.g. written by old client). Nothing to validate.
    // 3) OR: The Stored checksum matches the calculated one. Validation succeeded.
    JsonUtils.fromJson[LastCheckpointInfo](jsonStr)
  }

  /**
   * Analyzes the json representation of [[LastCheckpointInfo]] and returns checksum tuple where
   * - first element refers to the stored checksum in the json representation of
   *   [[LastCheckpointInfo]], None if the checksum is not present.
   * - second element refers to the checksum computed from the canonicalized json representation of
   *   the [[LastCheckpointInfo]].
   */
  def getChecksums(jsonStr: String): (Option[String], String) = {
    val reader =
      JsonUtils.mapper.reader().withFeatures(DeserializationFeature.FAIL_ON_READING_DUP_TREE_KEY)
    val rootNode = reader.readTree(jsonStr)
    val storedChecksum = if (rootNode.has(STORED_CHECKSUM_KEY)) {
      Some(rootNode.get(STORED_CHECKSUM_KEY).asText())
    } else {
      None
    }
    val actualChecksum = treeNodeToChecksum(rootNode)
    storedChecksum -> actualChecksum
  }

  /**
   * Canonicalizes the given `treeNode` json and returns its md5 checksum.
   * Refer to "JSON checksum" section in PROTOCOL.md for canonicalization steps.
   */
  def treeNodeToChecksum(treeNode: JsonNode): String = {
    val jsonEntriesBuffer = ArrayBuffer.empty[(String, String)]

    import scala.collection.JavaConverters._
    def traverseJsonNode(currentNode: JsonNode, prefix: ArrayBuffer[String]): Unit = {
      if (currentNode.isObject) {
        currentNode.fields().asScala.foreach { entry =>
          prefix.append(encodeString(entry.getKey))
          traverseJsonNode(entry.getValue, prefix)
          prefix.trimEnd(1)
        }
      } else if (currentNode.isArray) {
        currentNode.asScala.zipWithIndex.foreach { case (jsonNode, index) =>
          prefix.append(index.toString)
          traverseJsonNode(jsonNode, prefix)
          prefix.trimEnd(1)
        }
      } else {
        var nodeValue = currentNode.asText()
        if (currentNode.isTextual) nodeValue = encodeString(nodeValue)
        jsonEntriesBuffer.append(prefix.mkString("+") -> nodeValue)
      }
    }
    traverseJsonNode(treeNode, prefix = ArrayBuffer.empty)
    import Ordering.Implicits._
    val normalizedJsonKeyValues = jsonEntriesBuffer
      .filter { case (k, _) => k != s""""$STORED_CHECKSUM_KEY"""" }
      .map { case (k, v) => s"$k=$v" }
      .sortBy(_.toSeq: Seq[Char])
      .mkString(",")
    DigestUtils.md5Hex(normalizedJsonKeyValues)
  }

  private val isUnreservedOctet =
    (Set.empty ++ ('a' to 'z') ++ ('A' to 'Z') ++ ('0' to '9') ++ "-._~").map(_.toByte)

  /**
   * URL encodes a String based on the following rules:
   * 1. Use uppercase hexadecimals for all percent encodings
   * 2. percent-encode everything other than unreserved characters
   * 3. unreserved characters are = a-z / A-Z / 0-9 / "-" / "." / "_" / "~"
   */
  private def encodeString(str: String): String = {
    val result = str.getBytes(java.nio.charset.StandardCharsets.UTF_8).map {
      case b if isUnreservedOctet(b) => b.toChar.toString
      case b =>
        // convert to char equivalent of unsigned byte
        val c = (b & 0xff)
        f"%%$c%02X"
    }.mkString
    s""""$result""""
  }

  def fromFiles(files: Seq[FileStatus]): LastCheckpointInfo = {
    assert(files.nonEmpty, "files should be non empty to construct LastCheckpointInfo")
    LastCheckpointInfo(
      version = checkpointVersion(files.head),
      size = -1L,
      parts = numCheckpointParts(files.head.getPath),
      sizeInBytes = Some(files.map(_.getLen).sum),
      numOfAddFiles = None,
      checkpointSchema = None
    )
  }
}

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/checkpoints/LastCheckpointInfo.java
/*
 * Copyright (2026) The Delta Lake Project Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.delta.kernel.internal.checkpoints;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import io.delta.kernel.internal.util.JsonUtils;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;

/**
 * Parsed, typed view of a {@code _last_checkpoint} file.
 *
 * <p>{@link CheckpointMetaData} projects only the classic columnar fields ({@code version}, {@code
 * size}, {@code parts}, {@code tags}) because the columnar JSON reader cannot express the on-disk
 * pointer's optional and recursive fields (notably {@code checkpointSchema}, a schema-of-schema).
 * This type is the counterpart for callers that need the full pointer -- e.g. the SoftStore /
 * metadata-cache path that mirrors these fields into a wire format. It is produced from the raw
 * blob (see {@link LastCheckpointSerialized}) via a plain JSON tree walk, so no field has to be
 * projectable into a fixed columnar schema.
 *
 * <p>The field set mirrors DBR's {@code LastCheckpointInfo} case class one-for-one, except for the
 * Edge-only {@code checkpointFiles} / {@code consecutiveCompactedFiles}, which are not JSON
 * serialized into {@code _last_checkpoint} and so are never present in the blob.
 *
 * <p>{@code checkpointSchema} is deliberately kept as its raw JSON string ({@code StructType.json})
 * rather than decoded into a {@code StructType}: it round-trips losslessly and semantically, and it
 * avoids eagerly decoding a recursive schema-of-schema that most callers relay verbatim. Callers
 * that want the structural type can parse it with {@code DataTypeJsonSerDe.deserializeStructType}.
 */
public final class LastCheckpointInfo {

  // ----- _last_checkpoint field names (must match DBR LastCheckpointInfo's JSON) -----
  private static final String VERSION = "version";
  private static final String SIZE = "size";
  private static final String PARTS = "parts";
  private static final String SIZE_IN_BYTES = "sizeInBytes";
  private static final String NUM_OF_ADD_FILES = "numOfAddFiles";
  private static final String CHECKPOINT_SCHEMA = "checkpointSchema";
  private static final String CHECKSUM = "checksum";
  private static final String V2_CHECKPOINT = "v2Checkpoint";

  /**
   * Parses a {@code _last_checkpoint} JSON blob into a {@link LastCheckpointInfo}.
   *
   * @param json the raw {@code _last_checkpoint} contents (see {@link
   *     LastCheckpointSerialized#json()})
   * @return the parsed pointer
   * @throws IllegalArgumentException if the blob is not a JSON object or lacks the required {@code
   *     version} / {@code size} fields
   */
  public static LastCheckpointInfo fromJson(String json) {
    JsonNode root;
    try {
      root = JsonUtils.mapper().readTree(json);
    } catch (Exception e) {
      throw new IllegalArgumentException("_last_checkpoint is not parseable JSON", e);
    }
    return fromJsonNode(root);
  }

  /** Parses an already-decoded {@code _last_checkpoint} tree. See {@link #fromJson(String)}. */
  public static LastCheckpointInfo fromJsonNode(JsonNode root) {
    if (root == null || !root.isObject()) {
      throw new IllegalArgumentException("_last_checkpoint must be a JSON object");
    }
    return new LastCheckpointInfo(
        requiredLong(root, VERSION),
        requiredLong(root, SIZE),
        optionalInt(root, PARTS),
        optionalLong(root, SIZE_IN_BYTES),
        optionalLong(root, NUM_OF_ADD_FILES),
        // Keep the schema-of-schema as its raw JSON string; see class doc.
        rawJson(root, CHECKPOINT_SCHEMA),
        optionalString(root, CHECKSUM),
        LastCheckpointV2.fromJsonNode(root.get(V2_CHECKPOINT)));
  }

  private final long version;
  private final long size;
  private final Optional<Integer> parts;
  private final Optional<Long> sizeInBytes;
  private final Optional<Long> numOfAddFiles;
  private final Optional<String> checkpointSchemaJson;
  private final Optional<String> checksum;
  private final Optional<LastCheckpointV2> v2Checkpoint;

  public LastCheckpointInfo(
      long version,
      long size,
      Optional<Integer> parts,
      Optional<Long> sizeInBytes,
      Optional<Long> numOfAddFiles,
      Optional<String> checkpointSchemaJson,
      Optional<String> checksum,
      Optional<LastCheckpointV2> v2Checkpoint) {
    this.version = version;
    this.size = size;
    this.parts = parts;
    this.sizeInBytes = sizeInBytes;
    this.numOfAddFiles = numOfAddFiles;
    this.checkpointSchemaJson = checkpointSchemaJson;
    this.checksum = checksum;
    this.v2Checkpoint = v2Checkpoint;
  }

  public long getVersion() {
    return version;
  }

  public long getSize() {
    return size;
  }

  public Optional<Integer> getParts() {
    return parts;
  }

  public Optional<Long> getSizeInBytes() {
    return sizeInBytes;
  }

  public Optional<Long> getNumOfAddFiles() {
    return numOfAddFiles;
  }

  /**
   * The checkpoint schema as its raw {@code StructType.json} string, or empty when the pointer
   * omits it. Not decoded into a {@code StructType}; see the class doc for why.
   */
  public Optional<String> getCheckpointSchemaJson() {
    return checkpointSchemaJson;
  }

  public Optional<String> getChecksum() {
    return checksum;
  }

  public Optional<LastCheckpointV2> getV2Checkpoint() {
    return v2Checkpoint;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (!(o instanceof LastCheckpointInfo)) {
      return false;
    }
    LastCheckpointInfo that = (LastCheckpointInfo) o;
    return version == that.version
        && size == that.size
        && parts.equals(that.parts)
        && sizeInBytes.equals(that.sizeInBytes)
        && numOfAddFiles.equals(that.numOfAddFiles)
        && checkpointSchemaJson.equals(that.checkpointSchemaJson)
        && checksum.equals(that.checksum)
        && v2Checkpoint.equals(that.v2Checkpoint);
  }

  @Override
  public int hashCode() {
    return Objects.hash(
        version,
        size,
        parts,
        sizeInBytes,
        numOfAddFiles,
        checkpointSchemaJson,
        checksum,
        v2Checkpoint);
  }

  @Override
  public String toString() {
    return "LastCheckpointInfo{"
        + "version="
        + version
        + ", size="
        + size
        + ", parts="
        + parts
        + ", sizeInBytes="
        + sizeInBytes
        + ", numOfAddFiles="
        + numOfAddFiles
        + ", checkpointSchemaJson="
        + (checkpointSchemaJson.isPresent() ? "<present>" : "<absent>")
        + ", checksum="
        + checksum
        + ", v2Checkpoint="
        + v2Checkpoint
        + '}';
  }

  /**
   * The V2-checkpoint pointer nested under {@code v2Checkpoint}. Mirrors DBR's {@code
   * LastCheckpointV2}.
   *
   * <p>{@code nonFileActions} is exposed as the list of verbatim per-action JSON objects (e.g.
   * {@code {"protocol":{...}}}, {@code {"metaData":{...}}}, {@code {"checkpointMetadata":{...}}}).
   * Decoding those into fully typed action objects is intentionally left to the consumer: the
   * action model (Metadata / Protocol / DomainMetadata / ...) is large, and the SoftStore codec
   * maps them into wire types the kernel has no knowledge of. The single action every consumer
   * needs -- {@code checkpointMetadata} -- is surfaced directly via {@link
   * #getCheckpointMetadataVersion()}. {@code sidecarFiles} is fully parsed since it is small.
   */
  public static final class LastCheckpointV2 {
    private static final String PATH = "path";
    private static final String SIZE_IN_BYTES = "sizeInBytes";
    private static final String MODIFICATION_TIME = "modificationTime";
    private static final String NON_FILE_ACTIONS = "nonFileActions";
    private static final String SIDECAR_FILES = "sidecarFiles";
    private static final String CHECKPOINT_METADATA = "checkpointMetadata";
    private static final String VERSION = "version";

    static Optional<LastCheckpointV2> fromJsonNode(JsonNode node) {
      if (node == null || !node.isObject()) {
        return Optional.empty();
      }
      return Optional.of(
          new LastCheckpointV2(
              requiredString(node, PATH),
              requiredLong(node, SIZE_IN_BYTES),
              requiredLong(node, MODIFICATION_TIME),
              rawJsonArrayElements(node.get(NON_FILE_ACTIONS)),
              SidecarFileInfo.fromArray(node.get(SIDECAR_FILES))));
    }

    private final String path;
    private final long sizeInBytes;
    private final long modificationTime;
    // Absent (None) vs empty list is preserved: absent = info missing, empty = no such actions.
    private final Optional<List<String>> nonFileActionsJson;
    private final Optional<List<SidecarFileInfo>> sidecarFiles;

    public LastCheckpointV2(
        String path,
        long sizeInBytes,
        long modificationTime,
        Optional<List<String>> nonFileActionsJson,
        Optional<List<SidecarFileInfo>> sidecarFiles) {
      this.path = path;
      this.sizeInBytes = sizeInBytes;
      this.modificationTime = modificationTime;
      this.nonFileActionsJson = nonFileActionsJson;
      this.sidecarFiles = sidecarFiles;
    }

    public String getPath() {
      return path;
    }

    public long getSizeInBytes() {
      return sizeInBytes;
    }

    public long getModificationTime() {
      return modificationTime;
    }

    /** The non-file actions as verbatim per-action JSON objects. See {@link LastCheckpointV2}. */
    public Optional<List<String>> getNonFileActionsJson() {
      return nonFileActionsJson;
    }

    public Optional<List<SidecarFileInfo>> getSidecarFiles() {
      return sidecarFiles;
    }

    /**
     * The {@code version} of the {@code checkpointMetadata} action among {@code nonFileActions}, or
     * empty when the pointer carries no such action. This is the field the metadata-cache path uses
     * to decide whether a V2 pointer is cacheable.
     */
    public Optional<Long> getCheckpointMetadataVersion() {
      if (!nonFileActionsJson.isPresent()) {
        return Optional.empty();
      }
      for (String actionJson : nonFileActionsJson.get()) {
        JsonNode action;
        try {
          action = JsonUtils.mapper().readTree(actionJson);
        } catch (Exception e) {
          continue;
        }
        JsonNode cm = action.get(CHECKPOINT_METADATA);
        if (cm != null && cm.isObject()) {
          return optionalLong(cm, VERSION);
        }
      }
      return Optional.empty();
    }

    @Override
    public boolean equals(Object o) {
      if (this == o) {
        return true;
      }
      if (!(o instanceof LastCheckpointV2)) {
        return false;
      }
      LastCheckpointV2 that = (LastCheckpointV2) o;
      return sizeInBytes == that.sizeInBytes
          && modificationTime == that.modificationTime
          && path.equals(that.path)
          && nonFileActionsJson.equals(that.nonFileActionsJson)
          && sidecarFiles.equals(that.sidecarFiles);
    }

    @Override
    public int hashCode() {
      return Objects.hash(path, sizeInBytes, modificationTime, nonFileActionsJson, sidecarFiles);
    }

    @Override
    public String toString() {
      return "LastCheckpointV2{path=" + path + ", sizeInBytes=" + sizeInBytes + '}';
    }
  }

  /** A sidecar-file pointer inside a V2 checkpoint. Mirrors DBR's {@code SidecarFile}. */
  public static final class SidecarFileInfo {
    private static final String PATH = "path";
    private static final String SIZE_IN_BYTES = "sizeInBytes";
    private static final String MODIFICATION_TIME = "modificationTime";
    private static final String TAGS = "tags";

    static Optional<List<SidecarFileInfo>> fromArray(JsonNode arrayNode) {
      if (arrayNode == null || !arrayNode.isArray()) {
        return Optional.empty();
      }
      List<SidecarFileInfo> files = new ArrayList<>();
      for (JsonNode f : arrayNode) {
        files.add(
            new SidecarFileInfo(
                requiredString(f, PATH),
                requiredLong(f, SIZE_IN_BYTES),
                requiredLong(f, MODIFICATION_TIME),
                stringMap(f.get(TAGS))));
      }
      return Optional.of(Collections.unmodifiableList(files));
    }

    private final String path;
    private final long sizeInBytes;
    private final long modificationTime;
    private final Map<String, String> tags;

    public SidecarFileInfo(
        String path, long sizeInBytes, long modificationTime, Map<String, String> tags) {
      this.path = path;
      this.sizeInBytes = sizeInBytes;
      this.modificationTime = modificationTime;
      this.tags = tags;
    }

    public String getPath() {
      return path;
    }

    public long getSizeInBytes() {
      return sizeInBytes;
    }

    public long getModificationTime() {
      return modificationTime;
    }

    public Map<String, String> getTags() {
      return tags;
    }

    @Override
    public boolean equals(Object o) {
      if (this == o) {
        return true;
      }
      if (!(o instanceof SidecarFileInfo)) {
        return false;
      }
      SidecarFileInfo that = (SidecarFileInfo) o;
      return sizeInBytes == that.sizeInBytes
          && modificationTime == that.modificationTime
          && path.equals(that.path)
          && tags.equals(that.tags);
    }

    @Override
    public int hashCode() {
      return Objects.hash(path, sizeInBytes, modificationTime, tags);
    }

    @Override
    public String toString() {
      return "SidecarFileInfo{path=" + path + ", sizeInBytes=" + sizeInBytes + '}';
    }
  }

  // ----- JSON helpers -----

  private static long requiredLong(JsonNode node, String field) {
    JsonNode v = node.get(field);
    if (v == null || !v.isNumber()) {
      throw new IllegalArgumentException(
          "_last_checkpoint is missing required numeric field: " + field);
    }
    return v.asLong();
  }

  private static String requiredString(JsonNode node, String field) {
    JsonNode v = node.get(field);
    if (v == null || !v.isTextual()) {
      throw new IllegalArgumentException(
          "_last_checkpoint is missing required string field: " + field);
    }
    return v.asText();
  }

  private static Optional<Long> optionalLong(JsonNode node, String field) {
    JsonNode v = node.get(field);
    return (v != null && v.isNumber()) ? Optional.of(v.asLong()) : Optional.empty();
  }

  private static Optional<Integer> optionalInt(JsonNode node, String field) {
    JsonNode v = node.get(field);
    return (v != null && v.isNumber()) ? Optional.of(v.asInt()) : Optional.empty();
  }

  private static Optional<String> optionalString(JsonNode node, String field) {
    JsonNode v = node.get(field);
    return (v != null && v.isTextual()) ? Optional.of(v.asText()) : Optional.empty();
  }

  /** Returns the compact JSON text of a sub-object field (e.g. {@code checkpointSchema}). */
  private static Optional<String> rawJson(JsonNode node, String field) {
    JsonNode v = node.get(field);
    return (v != null && !v.isNull()) ? Optional.of(v.toString()) : Optional.empty();
  }

  /** Returns each element of a JSON array as its verbatim JSON text; empty Optional if absent. */
  private static Optional<List<String>> rawJsonArrayElements(JsonNode arrayNode) {
    if (arrayNode == null || !arrayNode.isArray()) {
      return Optional.empty();
    }
    List<String> out = new ArrayList<>();
    for (JsonNode e : arrayNode) {
      out.add(e.toString());
    }
    return Optional.of(Collections.unmodifiableList(out));
  }

  private static Map<String, String> stringMap(JsonNode node) {
    if (node == null || !node.isObject()) {
      return Collections.emptyMap();
    }
    Map<String, String> map = new LinkedHashMap<>();
    ObjectNode obj = (ObjectNode) node;
    obj.fields().forEachRemaining(e -> map.put(e.getKey(), e.getValue().asText()));
    return Collections.unmodifiableMap(map);
  }
}
